iac-extensions/terraform/src/main/java/org/sonar/iac/terraform/checks/ResourceVisitor.java

```diff
@@ -0,0 +1,223 @@
+
+  @Override
+  public void initialize(InitContext init) {
+    init.register(BlockTree.class, this::provideResource);
+    registerResourceConsumer();
+  }
+
+  protected abstract void registerResourceConsumer();
+
+  protected void provideResource(CheckContext ctx, BlockTree blockTree) {
+    if (isResource(blockTree)) {
+      Resource resource = new Resource(ctx, blockTree);
+      if (resourceConsumers.containsKey(resource.type)) {
+        resourceConsumers.get(resource.type).forEach(consumer -> consumer.accept(resource));
+      }
+    }
+  }
+
+  protected void register(String resourceName, Consumer<Resource> consumer) {
+    resourceConsumers.computeIfAbsent(resourceName, i -> new ArrayList<>()).add(consumer);
+  }
+
+  protected void register(List<String> resourceNames, Consumer<Resource> consumer) {
+    resourceNames.forEach(resourceName -> register(resourceName, consumer));
+  }
+
+  protected static class Block {
+
+    protected final CheckContext ctx;
+    protected final BlockTree blockTree;
+
+    private Block(CheckContext ctx, BlockTree blockTree) {
+      this.ctx = ctx;
+      this.blockTree = blockTree;
+    }
+
+    public Attribute attribute(String propertyName) {
+      return PropertyUtils.get(blockTree, propertyName, AttributeTree.class)
+        .map(a -> (Attribute) new Attribute.PresentAttribute(ctx, this, propertyName, a))
+        .orElse(new Attribute.AbsentAttribute(ctx, this, propertyName));
+    }
+
+    public Optional<Block> block(String propertyName) {
+      return PropertyUtils.get(blockTree, propertyName, BlockTree.class).map(b -> new Block(ctx, b));
+    }
+
+    public Stream<Block> blocks(String propertyName) {
+      return PropertyUtils.getAll(blockTree, propertyName, BlockTree.class).stream().map(b -> new Block(ctx, b));
+    }
+
+    public void report(String message, SecondaryLocation... secondaries) {
+      ctx.reportIssue(blockTree.key(), message, Arrays.asList(secondaries));
+    }
+  }
+
+  protected static class Resource extends Block  {
+    private final String type;
+
+    public Resource(CheckContext ctx, BlockTree resourceTree) {
+      super(ctx, resourceTree);
+      this.type = getResourceType(resourceTree);
+    }
+
+    @Override
+    public void report(String message, SecondaryLocation... secondaries) {
+      ctx.reportIssue(blockTree.labels().get(0), message, Arrays.asList(secondaries));
+    }
+  }
+
+  public static class Attribute {
+
+    protected final CheckContext ctx;
+    protected final Block block;
+    protected final String name;
+    @Nullable
+    protected final AttributeTree attributeTree;
+
+    public Attribute(CheckContext ctx, Block block, String name, @Nullable AttributeTree attributeTree) {
+      this.ctx = ctx;
+      this.block = block;
+      this.name = name;
+      this.attributeTree = attributeTree;
+    }
+
+    public Attribute reportOnTrue(String message, SecondaryLocation... secondaries) {
```

Comment:
What about `reportIfTrue`/`reportIfFalse`/`reportIfAbsent` for consistency with the other methods which were just renamed? I feel it might also clarify what the method actually do.

Reply:
Ok for me